[SE-327] Implement flow-sensitive actor isolation for actor inits

kavon opened the pull request:
This replaces the escaping-use restriction, expanding the set of programs accepted by the compiler _without_ warnings. Specifically, this PR fixes:

- using data-flow analysis, instead of an escaping-use restriction, to prove actor initializers are race-free (rdar://84476555)
  - illegal property accesses detected by flow-isolation are diagnosed as warnings until Swift 6.
- a missing hop-to-executor (in the prologue) for global-actor isolated, async, actor-type inits. (rdar://87412308)
- removes the Sendable warnings when assigning to a non-sendable stored property in an actor initializer (rdar://83985890)
- In Swift 6, 'defer' now correctly respects its context when deciding its property access kind. This makes a difference particularly in initializers.

This PR implements many but not all of the pieces of SE-327:
- Flow-sensitive actor isolation checking.
- Sendable checking only during calls to actor initializers, instead of when storing to non-sendable properties of the actor from the init.

This PR also has a known issue with delegating async initializers in an actor, where a hop is missing after calls to async functions. I decided that this PR is getting too big and that issue will be solved in a follow-up PR.

kavon committed:
basic implementation of flow-isolation for SE-327

Flow-isolation is a diagnostic SIL pass that finds
unsafe accesses to properties in initializers and
deinitializers that cannot gain isolation to otherwise
protect those accesses from concurrent modifications.
See SE-327 for more details about how and why it exists.

This commit includes changes and features like:

- The removal of the escaping-use restriction
- Flow-isolation that works properly with `defer` statements
- Flow-isolation with an emphasis on helpful diagnostics.

It also includes known issues like:

- Local / nonescaping functions are not analyzed by
  flow-isolation, despite it being technically possible.
  The main challenge in supporting it efficiently is that
  such functions do not have a single exit-point, like
  a `defer`. In particular, arbitrary functions can throw
  so there are points where nonisolation should _not_ flow
  out of the function at a call-site in the initializer, etc.

- The implementation of the flow-isolation pass is not
  particularly memory efficient; it relies on BitDataflow
  even though the particular flow problem is simple.
  So, a more efficient implementation would be specialized for
  this particular problem, etc.

There are also some changes to the Swift language itself: defer
will respect its context when deciding its property access kind.

Previously, a defer in an initializer would always access a stored
property through its accessor methods, instead of doing so directly
like its enclosing function might. This inconsistency is unfortunate,
so for Swift 6+ we make this consistent. For Swift 5, only a defer
in a function that is a member of the following kinds of types
will gain this consistency:

- an actor type
- any nominal type that is actor-isolated, excluding UnsafeGlobalActor.

These types are still rather new, so there is much less of a chance of
breaking expected behaviors around defer. In particular, the danger is
that users are relying on the behavior of defer triggering a property
observer within an init or deinit, when it would not be triggering it
without the defer.

atrick commented on lib/SILOptimizer/Mandatory/FlowIsolation.cpp:
```diff
@@ -0,0 +1,608 @@
+    switch (user->getKind()) {
+      // isolated uses
+      case SILInstructionKind::StoreInst:
+      case SILInstructionKind::LoadInst:
```
Formal access should always have an access scope. I'm curious why we need to look for individual load/store instructions. That doesn't see, reliable.

kavon replied:
Thanks, I hadn't thought of that. I've fixed this and now it uses the access scope, so it can handle other kinds of accesses. I might need to refine this a bit more to hang onto the `ref_element_addr` so that diagnostics underline the right thing. Sometimes it underlines the `=` in an assignment, etc.

kavon replied:
I had to go with tracking RefElementAddrs, because not all initializers used formal access scopes correctly (I can't recall which ones unfortunately).

atrick commented on lib/SILOptimizer/Mandatory/FlowIsolation.cpp:
```diff
@@ -0,0 +1,608 @@
+                        std::unique_ptr<AnalysisInfo>, 8> deferBlocks;
+
+  // Only computed after calling solve()
+  BitDataflow flow;
```
This is using 4 bitvectors per basic block to analyze the uses of a single SSA value! It can all be done with just two BasicBlockSets for the entire function. One for the isolated blocks and one for the nonisolated blocks.

kavon replied:
Hm, yea. I might be able to implement this even more efficiently than that, too. 

I think the only information I truly need is:
1. whether a block generates nonisolation, which is already tracked with the `nonisolatedUses` set in each block's corresponding `Info`.
2. whether the block exits with nonisolation, which can be one BasicBlockSet for the whole function.

Once I've got a bunch more tests in place to make sure it's working the way we need, I'll go ahead and rewrite it to be more memory efficient, dropping the need for BitDataflow.

kavon committed:
(rdar://87412308) fix missing actor hop for global-actor isolated inits of an actor type.

For global-actor isolated async inits within an actor, we will have separate restrictions
on access to self's differently-isolated stored properties to prevent data races. But,
we also must be on the global actor while in such initializers. The problem was only
that the prologue was missing a hop because of a bad conditional check. Previously
the check assumed that actors cannot have global-actor isolated initializers, but they
can currently.

kavon committed:
fix bugs with actor inits and flow-isolation

This is a combination of fixes:

- inject hops in self-isolated delegating actor initializers
  after self becomes initialized.

- fix sendable and isolation for convenience inits

- fix bug in distributed actor inits that I introduced when
  implementing flow-isolation.

- fix / add test coverage.

kavon commented:
@swift-ci please test linux platform

kavon committed:
update observers test to require concurrency

kavon commented:
@swift-ci please test

kavon committed:
allow nonisolated + async + delegating actor inits.

kavon commented:
@swift-ci please test

kavon committed:
cleanup

kavon commented:
@swift-ci please test

kavon committed:
test updates after rebasing on main

these updates primarily are to account for the new generic executor
that is being used for nonisolated / unspecified async functions.

kavon commented:
@swift-ci please test and merge

kavon commented:
Hmm, getting this when building on Linux:

```
13:36:47 FAILED: lib/SILOptimizer/CMakeFiles/swiftSILOptimizer.dir/Mandatory/FlowIsolation.cpp.o 
13:36:47 /usr/bin/clang++ -DCMARK_STATIC_DEFINE -DGTEST_HAS_RTTI=0 -DSWIFT_LLVM_SUPPORT_IS_AVAILABLE -D_DEBUG -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -Ilib/SILOptimizer -I/home/build-user/swift/lib/SILOptimizer -Iinclude -I/home/build-user/swift/include -I/home/build-user/llvm-project/llvm/include -I/home/build-user/build/buildbot_linux/llvm-linux-x86_64/include -I/home/build-user/llvm-project/clang/include -I/home/build-user/build/buildbot_linux/llvm-linux-x86_64/tools/clang/include -I/home/build-user/cmark/src -I/home/build-user/build/buildbot_linux/cmark-linux-x86_64/src -I/home/build-user/swift-corelibs-libdispatch/src/BlocksRuntime -I/home/build-user/swift-corelibs-libdispatch -Wno-unknown-warning-option -Werror=unguarded-availability-new -fno-stack-protector -fPIC -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -Wimplicit-fallthrough -Wcovered-switch-default -Wno-class-memaccess -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wstring-conversion -Wmisleading-indentation -fdiagnostics-color -ffunction-sections -fdata-sections -Werror=switch -Wdocumentation -Wimplicit-fallthrough -Wunreachable-code -Woverloaded-virtual -DOBJC_OLD_DISPATCH_PROTOTYPES=0 -O3 -DNDEBUG  -fno-exceptions -fno-rtti -Werror=gnu -Werror=c++98-compat-extra-semi -UNDEBUG -target x86_64-unknown-linux-gnu -O2 -g0 -march=core2 -std=c++14 -MD -MT lib/SILOptimizer/CMakeFiles/swiftSILOptimizer.dir/Mandatory/FlowIsolation.cpp.o -MF lib/SILOptimizer/CMakeFiles/swiftSILOptimizer.dir/Mandatory/FlowIsolation.cpp.o.d -o lib/SILOptimizer/CMakeFiles/swiftSILOptimizer.dir/Mandatory/FlowIsolation.cpp.o -c /home/build-user/swift/lib/SILOptimizer/Mandatory/FlowIsolation.cpp
13:36:47 In file included from /home/build-user/swift/lib/SILOptimizer/Mandatory/FlowIsolation.cpp:15:
13:36:47 In file included from /home/build-user/swift/include/swift/AST/ActorIsolation.h:19:
13:36:47 In file included from /home/build-user/swift/include/swift/AST/Type.h:23:
13:36:47 In file included from /home/build-user/llvm-project/llvm/include/llvm/ADT/DenseMap.h:24:
13:36:47 In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/algorithm:62:
13:36:47 
/usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/stl_algo.h:162:10: error: no matching function for call to '__iterator_category'
13:36:47                        std::__iterator_category(__first));
13:36:47                        ^~~~~~~~~~~~~~~~~~~~~~~~
13:36:47 /usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/stl_algo.h:3963:19: note: in instantiation of function template specialization 'std::__find_if<swift::BasicBlockData<(anonymous namespace)::Info, 32>::IteratorTy<llvm::ilist_iterator<llvm::ilist_detail::node_options<swift::SILBasicBlock, true, false, void>, false, false>, true>, __gnu_cxx::__ops::_Iter_pred<(lambda at /home/build-user/swift/lib/SILOptimizer/Mandatory/FlowIsolation.cpp:191:40)> >' requested here
13:36:47       return std::__find_if(__first, __last,
13:36:47                   ^
13:36:47 /usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/stl_algo.h:526:40: note: in instantiation of function template specialization 'std::find_if<swift::BasicBlockData<(anonymous namespace)::Info, 32>::IteratorTy<llvm::ilist_iterator<llvm::ilist_detail::node_options<swift::SILBasicBlock, true, false, void>, false, false>, true>, (lambda at /home/build-user/swift/lib/SILOptimizer/Mandatory/FlowIsolation.cpp:191:40)>' requested here
13:36:47     { return __last == _GLIBCXX_STD_A::find_if(__first, __last, __pred); }
13:36:47                                        ^
13:36:47 /usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/stl_algo.h:544:20: note: in instantiation of function template specialization 'std::none_of<swift::BasicBlockData<(anonymous namespace)::Info, 32>::IteratorTy<llvm::ilist_iterator<llvm::ilist_detail::node_options<swift::SILBasicBlock, true, false, void>, false, false>, true>, (lambda at /home/build-user/swift/lib/SILOptimizer/Mandatory/FlowIsolation.cpp:191:40)>' requested here
13:36:47     { return !std::none_of(__first, __last, __pred); }
13:36:47                    ^
13:36:47 /home/build-user/swift/lib/SILOptimizer/Mandatory/FlowIsolation.cpp:191:17: note: in instantiation of function template specialization 'std::any_of<swift::BasicBlockData<(anonymous namespace)::Info, 32>::IteratorTy<llvm::ilist_iterator<llvm::ilist_detail::node_options<swift::SILBasicBlock, true, false, void>, false, false>, true>, (lambda at /home/build-user/swift/lib/SILOptimizer/Mandatory/FlowIsolation.cpp:191:40)>' requested here
13:36:47     return std::any_of(begin(), end(), [](auto const& bnd) -> bool {
13:36:47                 ^
13:36:47 /usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/stl_iterator_base_types.h:205:5: note: candidate template ignored: substitution failure [with _Iter = swift::BasicBlockData<(anonymous namespace)::Info, 32>::IteratorTy<llvm::ilist_iterator<llvm::ilist_detail::node_options<swift::SILBasicBlock, true, false, void>, false, false>, true>]: no type named 'iterator_category' in 'std::iterator_traits<swift::BasicBlockData<(anonymous namespace)::Info, 32>::IteratorTy<llvm::ilist_iterator<llvm::ilist_detail::node_options<swift::SILBasicBlock, true, false, void>, false, false>, true> >'
13:36:47     __iterator_category(const _Iter&)
13:36:47     ^
13:36:47 In file included from /home/build-user/swift/lib/SILOptimizer/Mandatory/FlowIsolation.cpp:15:
13:36:47 In file included from /home/build-user/swift/include/swift/AST/ActorIsolation.h:19:
13:36:47 In file included from /home/build-user/swift/include/swift/AST/Type.h:23:
13:36:47 In file included from /home/build-user/llvm-project/llvm/include/llvm/ADT/DenseMap.h:24:
13:36:47 In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/algorithm:62:
13:36:47 
/usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/stl_algo.h:162:10: error: no matching function for call to '__iterator_category'
13:36:47                        std::__iterator_category(__first));
13:36:47                        ^~~~~~~~~~~~~~~~~~~~~~~~
13:36:47 /usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/stl_algo.h:3963:19: note: in instantiation of function template specialization 'std::__find_if<swift::BasicBlockData<(anonymous namespace)::Info, 32>::IteratorTy<llvm::ilist_iterator<llvm::ilist_detail::node_options<swift::SILBasicBlock, true, false, void>, false, false>, true>, __gnu_cxx::__ops::_Iter_pred<(lambda at /home/build-user/swift/lib/SILOptimizer/Mandatory/FlowIsolation.cpp:198:40)> >' requested here
13:36:47       return std::__find_if(__first, __last,
13:36:47                   ^
13:36:47 /usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/stl_algo.h:526:40: note: in instantiation of function template specialization 'std::find_if<swift::BasicBlockData<(anonymous namespace)::Info, 32>::IteratorTy<llvm::ilist_iterator<llvm::ilist_detail::node_options<swift::SILBasicBlock, true, false, void>, false, false>, true>, (lambda at /home/build-user/swift/lib/SILOptimizer/Mandatory/FlowIsolation.cpp:198:40)>' requested here
13:36:47     { return __last == _GLIBCXX_STD_A::find_if(__first, __last, __pred); }
13:36:47                                        ^
13:36:47 /usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/stl_algo.h:544:20: note: in instantiation of function template specialization 'std::none_of<swift::BasicBlockData<(anonymous namespace)::Info, 32>::IteratorTy<llvm::ilist_iterator<llvm::ilist_detail::node_options<swift::SILBasicBlock, true, false, void>, false, false>, true>, (lambda at /home/build-user/swift/lib/SILOptimizer/Mandatory/FlowIsolation.cpp:198:40)>' requested here
13:36:47     { return !std::none_of(__first, __last, __pred); }
13:36:47                    ^
13:36:47 /home/build-user/swift/lib/SILOptimizer/Mandatory/FlowIsolation.cpp:198:17: note: in instantiation of function template specialization 'std::any_of<swift::BasicBlockData<(anonymous namespace)::Info, 32>::IteratorTy<llvm::ilist_iterator<llvm::ilist_detail::node_options<swift::SILBasicBlock, true, false, void>, false, false>, true>, (lambda at /home/build-user/swift/lib/SILOptimizer/Mandatory/FlowIsolation.cpp:198:40)>' requested here
13:36:47     return std::any_of(begin(), end(), [](auto const& bnd) -> bool {
13:36:47                 ^
13:36:47 /usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/stl_iterator_base_types.h:205:5: note: candidate template ignored: substitution failure [with _Iter = swift::BasicBlockData<(anonymous namespace)::Info, 32>::IteratorTy<llvm::ilist_iterator<llvm::ilist_detail::node_options<swift::SILBasicBlock, true, false, void>, false, false>, true>]: no type named 'iterator_category' in 'std::iterator_traits<swift::BasicBlockData<(anonymous namespace)::Info, 32>::IteratorTy<llvm::ilist_iterator<llvm::ilist_detail::node_options<swift::SILBasicBlock, true, false, void>, false, false>, true> >'
13:36:47     __iterator_category(const _Iter&)
13:36:47     ^
13:36:47 2 errors generated.
```

kavon commented:
@swift-ci please smoke test linux

kavon commented:
@swift-ci please test and merge

kastiglione commented:
I am getting errors in Linux builds for my PR too:

https://ci.swift.org/job/apple-llvm-project-pr-linux/3349/consoleFull#2449481723122a513-f36a-4c87-8ed7-cbc36a1ec144

The line in question:

https://github.com/apple/swift/blob/249ec7a5c825198ccd8fec89e95f4958462be239/lib/SILOptimizer/Mandatory/FlowIsolation.cpp#L388

Has this error:

```
lib/SILOptimizer/Mandatory/FlowIsolation.cpp:388:14: error: chosen constructor is explicit in copy-initialization
    return {"a call involving", "", ctx.Id_self};
           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
